controllers/vm_controller.go

```diff
@@ -0,0 +1,122 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"kubevirt.io/ssp-operator/pkg/monitoring/metrics"
+)
+
+// +kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
+
+const vmControllerName = "vm-controller"
+
+type VmReconciler struct {
+	client client.Client
+	log    logr.Logger
+}
+
+func CreateVmController(mgr ctrl.Manager) (*VmReconciler, error) {
+	return newVmReconciler(mgr)
+}
+
+func (r *VmReconciler) Name() string {
+	return vmControllerName
+}
+
+func (r *VmReconciler) Start(_ context.Context, mgr ctrl.Manager) error {
+	return r.setupController(mgr)
+}
+
+func newVmReconciler(mgr ctrl.Manager) (*VmReconciler, error) {
+	logger := ctrl.Log.WithName("controllers").WithName("VirtualMachines")
+	reconciler := &VmReconciler{
+		client: mgr.GetClient(),
+		log:    logger,
+	}
+
+	return reconciler, nil
+}
+
+func (r *VmReconciler) setupController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("vm-controller").
+		For(&kubevirtv1.VirtualMachine{}).
+		Complete(r)
+}
+
+func (r *VmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	r.log.V(1).Info("Starting vm reconciliation...", "request", req.String())
+	vm := kubevirtv1.VirtualMachine{}
+	err = r.client.Get(ctx, req.NamespacedName, &vm)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.log.Info("VM not found", "vm", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		r.log.V(1).Info("Error retrieving the VM", "vm", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	r.setVmVolumesMetrics(&vm)
+
+	return ctrl.Result{}, err
+}
+
+func (r *VmReconciler) setVmVolumesMetrics(vm *kubevirtv1.VirtualMachine) {
+	if vm.Spec.Template.Spec.Volumes == nil {
+		return
+	}
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+		pvc, err := r.getPVC(vm, volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			continue
```

Comment:
I think we are better off requeueing in case of HTTP errors, we don't want to just lose that piece of information

Reply:
and adding it to a list of errors, and continuing